include/pisa/query/queries.hpp

```diff
@@ -32,70 +26,112 @@ namespace pisa {
+    std::vector<term_id_type> terms;
+    std::vector<float> term_weights;
 };
 
-[[nodiscard]] inline auto parse_query(
-    std::string const &query_string,
-    TermProcessor process_term,
-    std::optional<std::unordered_set<term_id_type>> const &stopwords = std::nullopt) -> Query
+[[nodiscard]] auto split_query_at_colon(std::string const &query_string)
+    -> std::pair<std::optional<std::string>, std::string_view>
 {
-    std::optional<std::string> id = std::nullopt;
-    std::vector<term_id_type> parsed_query;
+    // query id : terms (or ids)
     auto colon = std::find(query_string.begin(), query_string.end(), ':');
+    std::optional<std::string> id;
     if (colon != query_string.end()) {
         id = std::string(query_string.begin(), colon);
     }
     auto pos = colon == query_string.end() ? query_string.begin() : std::next(colon);
     auto raw_query = std::string_view(&*pos, std::distance(pos, query_string.end()));
-    TermTokenizer tokenizer(raw_query);
+    return {std::move(id), std::move(raw_query)};
+}
+
+[[nodiscard]] auto parse_query_terms(std::string const &query_string, TermProcessor term_processor)
+    -> Query
+{
+    auto id_and_raw_query = split_query_at_colon(query_string);
+    TermTokenizer tokenizer(std::move(id_and_raw_query.second));
+    std::vector<term_id_type> parsed_query;
     for (auto term_iter = tokenizer.begin(); term_iter != tokenizer.end(); ++term_iter) {
         auto raw_term = *term_iter;
         try {
-            auto processed = process_term(std::string(raw_term));
-            if (processed) {
-                if (not stopwords or stopwords->find(*processed) == stopwords->end()) {
-                    parsed_query.push_back(std::move(*processed));
+            auto term = term_processor(raw_term);
+            if (term) {
+                if (!term_processor.is_stopword(*term)) {
+                    parsed_query.push_back(std::move(*term));
                 } else {
-                    spdlog::warn("Term `{}` is a stop word and will be ignored", *processed);
+                    spdlog::warn("Term `{}` is a stopword and will be ignored", *term);
                 }
             } else {
                 spdlog::warn("Term `{}` not found and will be ignored", raw_term);
             }
-        } catch (std::invalid_argument& err) {
+        } catch (std::invalid_argument &err) {
             spdlog::warn("Could not parse `{}` to a number", raw_term);
         }
     }
-    return {id, parsed_query, {}};
+    return {std::move(id_and_raw_query.first), std::move(parsed_query), {}};
+}
+
+[[nodiscard]] auto parse_query_ids(std::string const &query_string) -> Query
+{
+    auto id_and_raw_query = split_query_at_colon(query_string);
+    auto id = id_and_raw_query.first;
+    auto raw_query = id_and_raw_query.second;
+
+    std::vector<term_id_type> parsed_query;
+    std::vector<std::string> term_ids;
+    boost::split(term_ids, raw_query, boost::is_any_of("\t"));
```

Comment:
I was thinking, let's maybe split on any whitespace? I don't see a reason why this should be strictly a tab.